python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1675,3 +1679,64 @@ def asset_check_spec(self) -> AssetCheckSpec:
     @deprecated(**_get_deprecation_kwargs("log_event"))
     def log_event(self, event: UserEvent) -> None:
         return self._op_execution_context.log_event(event)
+
+
+def build_execution_context(
+    step_context: StepExecutionContext,
+) -> Union[OpExecutionContext, AssetExecutionContext]:
+    """Get the correct context based on the type of step (op or asset) and the user provided context
+    type annotation. Follows these rules.
+
+    step type     annotation                result
+    asset         AssetExecutionContext     AssetExecutionContext
+    asset         OpExecutionContext        OpExecutionContext
+    asset         None                      AssetExecutionContext
+    op            AssetExecutionContext     Error - we cannot init an AssetExecutionContext w/o an AssetsDefinition
+    op            OpExecutionContext        OpExecutionContext
+    op            None                      OpExecutionContext
+
+
+    For ops in graph-backed assets
+    step type     annotation                result
+    op            AssetExecutionContext     AssetExecutionContext
+    op            OpExecutionContext        OpExecutionContext
+    op            None                      OpExecutionContext
+
+    """
+    is_sda_step = step_context.is_sda_step
+    is_op_in_graph_asset = is_sda_step and step_context.is_op_in_graph
+    context_annotation = EmptyAnnotation
+    compute_fn = step_context.op_def._compute_fn  # noqa: SLF001
+    compute_fn = (
+        compute_fn
+        if isinstance(compute_fn, DecoratedOpFunction)
+        else DecoratedOpFunction(compute_fn)
+    )
+    if compute_fn.has_context_arg():
+        context_param = compute_fn.get_context_arg()
+        context_annotation = context_param.annotation
+
+    # TODO - i dont know how to move this check to Definition time since we don't know if the op is
```

Comment:
TODO - i'd like to move this check to one we can catch at definition time, but i'm not sure if we can since with just this 
```python
@op 
def my_op(context: AssetExecutionContext)
```
we don't know if the op will be used in a graph backed asset (meaning this annotation is fine) or in an op job (meaning this annotation is not fine)